libpod/container_api.go

```diff
@@ -593,6 +601,24 @@ func (c *Container) Cleanup(ctx context.Context) error {
 		return errors.Wrapf(ErrCtrStateInvalid, "container %s is running or paused, refusing to clean up", c.ID())
 	}
 
+	// If we have a restart policy match when updating the state, we need to
+	// restart the container.
+	// However, perform a full validation of restart policy first.
+	if c.state.RestartPolicyMatch {
+		if c.config.RestartPolicy == RestartPolicyOnFailure && c.state.ExitCode != 0 {
+			logrus.Debugf("Container %s restart policy trigger: on retry %d (of %d)",
+				c.ID(), c.state.RestartCount, c.config.RestartRetries)
+		}
+		if (c.config.RestartPolicy == RestartPolicyOnFailure && c.state.ExitCode != 0 &&
+			(c.config.RestartRetries > 0 && c.state.RestartCount < c.config.RestartRetries)) ||
+			c.config.RestartPolicy == RestartPolicyAlways {
+			// The container stopped. We need to restart it.
+			return c.handleRestartPolicy(ctx)
```

Comment:
@mheon, can you elaborate why chose to put the logic into Cleanup()? It took me a while to wrap my head around the fact we end up in a recursion here and wondered if we couldn't make the logic more explicit.

Reply:
`Cleanup()` is guaranteed to be invoked after the container exits by our cleanup process. I could have added an explicit API endpoint to handle restart policy, but that didn't seem proper - the API should hide this if it can, as it's meant to happen automatically.

The separation of the decision of whether to make the restart originally lived here as well, but the problem is that `Cleanup()` is guaranteed to fire, but we can't guarantee it fires first - especially with `podman run`, we immediately fire a `Wait()` after the container dies, which performs a state transition and ensures that the logic here would not fire.

Reply:
I also don't think we recurse here - handleRestartPolicy may call cleanup() on failure, but cleanup() is also called from here.

Reply:
Hm. It wouldn't hurt to move more of the restart logic into handleRestartPolicy itself, though

Reply:
Ah, I recall what the problem was, now.

`handleRestartPolicy()` wants to return immediately and not proceed down through `Cleanup()` if a restart was actually performed. With the logic in cleanup, we only called it when it was guaranteed to restart, which solved that.

I'll make it return a bool to indicate if a restart was actually performed.